include/util/packed_vector.hpp

```diff
@@ -31,188 +35,353 @@ inline void write(storage::io::FileWriter &writer,
+                                                         const PackedVector &vec);
+
+  private:
+    void allocate_blocks(std::size_t num_blocks)
     {
-        vec.push_back(last_elem);
+        vec.resize(vec.size() + num_blocks * BLOCK_WORDS);
     }
 
-    template <bool enabled = (Ownership == storage::Ownership::View)>
-    std::uint64_t vec_back(typename std::enable_if<enabled>::type * = nullptr)
+    inline InternalIndex get_internal_index(const std::size_t index) const
     {
-        return vec[cursor];
+        const auto block_offset = BLOCK_ELEMENTS * (index / BLOCK_ELEMENTS);
+        const std::uint8_t element_index = index % BLOCK_ELEMENTS;
+        const auto lower_word_index = block_offset + word_offset[element_index];
+
+        return InternalIndex{lower_word_index, element_index};
     }
 
-    template <bool enabled = (Ownership == storage::Ownership::View)>
-    std::uint64_t vec_back(typename std::enable_if<!enabled>::type * = nullptr)
+    inline T get_value(const InternalIndex internal_index) const
+    {
+        const auto lower_word = vec[internal_index.lower_word];
+        // note this can actually already be a word of the next block however in
+        // that case the upper mask will be 0.
+        // we make sure to have a sentinel element to avoid out-of-bounds errors.
+        const auto upper_word = vec[internal_index.lower_word + 1];
+        const auto value = get_lower_half_value<WordT, T>(lower_word,
+                                                          lower_mask[internal_index.element],
+                                                          lower_offset[internal_index.element]) |
+                           get_upper_half_value<WordT, T>(upper_word,
+                                                          upper_mask[internal_index.element],
+                                                          upper_offset[internal_index.element]);
+        return value;
+    }
+
+    inline void set_value(const InternalIndex internal_index, const T value)
+    {
+        auto &lower_word = vec[internal_index.lower_word];
+        auto &upper_word = vec[internal_index.lower_word + 1];
+        lower_word = set_lower_value<WordT, T>(lower_word,
+                                               lower_mask[internal_index.element],
+                                               lower_offset[internal_index.element],
+                                               value);
+        upper_word = set_upper_value<WordT, T>(upper_word,
+                                               upper_mask[internal_index.element],
+                                               upper_offset[internal_index.element],
+                                               value);
+    }
+
+    void initialize_mask()
     {
-        return vec.back();
+        // TODO port to constexptr function
+        const std::uint64_t mask = (1UL << Bits) - 1;
```

Comment:
[coliru does not support -m32](http://coliru.stacked-crooked.com/a/48abfe596ba3c184), but with `1ul` -m32 prints `1023`, `1ull` is ok, but maybe `std::uint64_t{1} << Bits`? so it can be asserted that `Bits < sizeof(std::uint64_t)` to avoid UB